From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Tue, 26 Nov 2024 20:16:35 -0800
Subject: [PATCH] demo with TrimPattern


diff --git a/src/main/java/io/papermc/paper/registry/data/TrimPatternRegistryEntry.java b/src/main/java/io/papermc/paper/registry/data/TrimPatternRegistryEntry.java
new file mode 100644
index 0000000000000000000000000000000000000000..bcba9393d8d6f6d19bad00b8c63fb68010b68d63
--- /dev/null
+++ b/src/main/java/io/papermc/paper/registry/data/TrimPatternRegistryEntry.java
@@ -0,0 +1,46 @@
+package io.papermc.paper.registry.data;
+
+import io.papermc.paper.registry.RegistryBuilder;
+import io.papermc.paper.registry.TypedKey;
+import net.kyori.adventure.key.Key;
+import net.kyori.adventure.text.Component;
+import org.bukkit.inventory.ItemType;
+import org.bukkit.inventory.meta.trim.TrimPattern;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Contract;
+
+@ApiStatus.Experimental
+@ApiStatus.NonExtendable
+public interface TrimPatternRegistryEntry {
+
+    @Contract(pure = true)
+    Key assetId();
+
+    @Contract(pure = true)
+    TypedKey<ItemType> templateItem();
+
+    @Contract(pure = true)
+    Component description();
+
+    @Contract(pure = true)
+    boolean decal();
+
+    @ApiStatus.Experimental
+    @ApiStatus.NonExtendable
+    interface Builder extends TrimPatternRegistryEntry, RegistryBuilder<TrimPattern> {
+
+        @Contract(value = "_ -> this", mutates = "this")
+        Builder assetId(Key assetId);
+
+        @Contract(value = "_ -> this", mutates = "this")
+        Builder templateItem(TypedKey<ItemType> templateItem);
+
+        @Contract(value = "_ -> this", mutates = "this")
+        Builder description(Component description);
+
+        @Contract(value = "_ -> this", mutates = "this")
+        Builder decal(boolean decal);
+    }
+
+
+}
